### PR TITLE
Support locales resources in native: countries, currencies, timezones, languages

### LINKED ISSUE #22808: Including resource bundles for different languages doesn't work
### Describe the bug

When running the following code, there are differences between Native mode and JVM mode:

```java
Log.fatal(Locale.ITALY.getDisplayCountry(Locale.ENGLISH));
Log.fatal(Locale.ITALY.getDisplayCountry(Locale.GERMAN));
```

JVM mode output:

```
FATAL [] (executor-thread-1) Italy
FATAL [] (executor-thread-1) Italien
```

Native mode output:

```
FATAL [] (executor-thread-1) Italy
FATAL [] (executor-thread-1) Italy
```

The "LocaleNames" bundle was included via a resources-config.json file:

```json
{
  "bundles": [
    { "name": "sun.util.resources.LocaleNames" }
  ]
}
```

This JSON file was passed via the additional build arguments:

`quarkus.native.additional-build-args=-H:ResourceConfigurationFiles=resources-config.json,-H:+IncludeAllLocales,-H:DefaultLocale=en`

The language-specific Quarkus configurations have also been set:
```
quarkus.locales=en,de
quarkus.default-locale=en-US
quarkus.native.user-country=US
quarkus.native.user-language=en
```

### Expected behavior

_No response_

### Actual behavior

_No response_

### How to Reproduce?

Small reproducer example: https://github.com/andlinger/quarkus-native-locales

### Output of `uname -a` or `ver`

Darwin Kernel Version 21.2.0: Sun Nov 28 20:28:54 PST 2021; root:xnu-8019.61.5~1/RELEASE_X86_64 x86_64

### Output of `java -version`

openjdk version "17.0.1" 2021-10-19 OpenJDK Runtime Environment GraalVM CE 21.3.0 (build 17.0.1+12-jvmci-21.3-b05) OpenJDK 64-Bit Server VM GraalVM CE 21.3.0 (build 17.0.1+12-jvmci-21.3-b05, mixed mode, sharing)

### GraalVM version (if different from Java)

_No response_

### Quarkus version or git rev

2.6.2.Final, 2.6.1.Final, 2.3.1.Final

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

------------------------------------------------------------ Gradle 7.2 ------------------------------------------------------------  Build time:   2021-08-17 09:59:03 UTC Revision:     a773786b58bb28710e3dc96c4d1a7063628952ad  Kotlin:       1.5.21 Groovy:       3.0.8 Ant:          Apache Ant(TM) version 1.10.9 compiled on September 27 2020 JVM:          17.0.1 (GraalVM Community 17.0.1+12-jvmci-21.3-b05) OS:           Mac OS X 12.1 x86_64

### Additional information

_No response_

### LINKED ISSUE #23338: [Native] Locales bundles, ServiceLoader flow for JDK; e.g. Frankreich vs. France
### Describe the bug

This is a followup on: :lady_beetle:  #22808 **Including resource bundles for different languages doesn't work**

# The problem
Unless `quarkus.native.auto-service-loader-registration=true` is used, i.e. `-H:+UseServiceLoaderFeature`, to trigger [ServiceLoaderFeature](https://github.com/oracle/graal/blob/release/graal-vm/22.0/substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/ServiceLoaderFeature.java#L98) in Graal, JDK Locale bundles, e.g. `sun.util.resources.cldr.ext.LocaleNames_de` are not loaded.

The code in [ServiceLoaderFeature](https://github.com/oracle/graal/blob/release/graal-vm/22.0/substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/ServiceLoaderFeature.java#L98) is missing finer grained include/exclude, so it bloats the image too much when used.

The result is that it silently falls back to `sun.util.resources.LocaleNames` English defaults. To illustrate:

```
System.out.printf("English: %s\n", Locale.FRANCE.getDisplayCountry(Locale.ENGLISH));
System.out.printf("German: %s\n", Locale.FRANCE.getDisplayCountry(Locale.GERMAN));
System.out.printf("Italian: %s\n", Locale.FRANCE.getDisplayCountry(Locale.ITALIAN));
```
Should print:
```
English: France
German: Frankreich
Italian: Francia
```
and not
```
English: France
German: France
Italian: France
```

# The solution (Quarkus) :grey_question: 
Naively deferring to `RuntimeInitializedClassBuildItem` or using `ReflectiveClassBuildItem` to init e.g. 
```
sun.util.resources.LocaleData$SupplementaryResourceBundleProvider
sun.util.resources.provider.SupplementaryLocaleDataProvider
sun.util.resources.LocaleData$CommonResourceBundleProvider
sun.util.resources.provider.LocaleDataProvider
```
is not enough as it doesn't help with the ServiceLoader architecture used to populate these implementations.
Looking into `sun.util.resources.Bundles` and discovering Quarkus' `ServiceProviderBuildItem` lead to attempts at something like:
```
services.produce(
        new ServiceProviderBuildItem("sun.util.resources.LocaleDataProvider",
                "sun.util.resources.LocaleData$CommonResourceBundleProvider"));
 services.produce(
        new ServiceProviderBuildItem("sun.util.resources.provider.SupplementaryLocaleDataProvider",
                "sun.util.resources.LocaleData$SupplementaryResourceBundleProvider"));

or...?

services.produce(
        new ServiceProviderBuildItem(
                 "sun.util.resources.LocaleData$CommonResourceBundleProvider",
                  "sun.util.resources.LocaleDataProvider"));
services.produce(
         new ServiceProviderBuildItem(
                 "sun.util.resources.LocaleData$SupplementaryResourceBundleProvider",
                  "sun.util.resources.provider.SupplementaryLocaleDataProvider"));
```
It does not help either.

Examining some extensions such as `liquibase-mongodb` shows me that ServiceLoaders are successfully runtime-inited in Native image elsewhere, although none of these usages touch JDK's own code.

WIP, I will keep digging around, building `core/deployment/src/main/java/io/quarkus/deployment/steps/LocaleProcessor.java` that would runtime init ServiceLaoders to produce desired locales. No luck so far.

Note that having this is required:
```
        resources.produce(new NativeImageResourceBundleBuildItem("sun.util.resources.LocaleNames"));
```
and yet doing anything like:
```
        resources.produce(new NativeImageResourceBundleBuildItem("sun.util.resources.ext.LocaleNames_de"));
        resources.produce(new NativeImageResourceBundleBuildItem("sun.util.resources.cldr.ext.LocaleNames_de"));
```
does not help.

# The solution (Graal) :grey_question: 
It might be also addressable on the Graal's side and apparently, someone already put some thought to it:
From  [ServiceLoaderFeature](https://github.com/oracle/graal/blob/release/graal-vm/22.0/substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/ServiceLoaderFeature.java#L98):

>         // TODO write a more sophisticated include/exclude filter to handle cases like GR-27605 ?

So perhaps, instead of registering hundreds of services via all ServiceLoaders found all over the place with the hammer like `-H:+UseServiceLoaderFeature`, I could patch it do something akin to: `-H:+UseServiceLoaderFeature=include:sun.util.resources.LocaleData$CommonResourceBundleProvider,sun.util.resources.LocaleData$SupplementaryResourceBundleProvider` to enable just what is apparently needed in this particular case?

Note that while we use countries' names to illustrate the issue here, it would be likely the same for the names of currencies etc.

### PR DESCRIPTION
**Pull request by Karm**
Fixes #23338
Fixes #22808

Adds logic to control GraalVM's `-H:IncludeLocales`.

~If user sets anything but US English for any of~

~quarkus.native.user-language~
~quarkus.native.user-country~

~or if user adds anything but US English to a new config option:~

~quarkus.native.locales~

~LoalesProcessor kicks in and augments native-image with necessary providers to enable translation~
~of countries names, timezones and currencies.~

~I intentionally didn't alter the option `quarkus.locales` that is used by Hibernate Validator.~
~It is converted and used for a different purpose than native-image runtime locales availability.~

See https://github.com/quarkusio/quarkus/pull/23518#issuecomment-1034382063

### CODE REVIEW

File: core/deployment/src/main/java/io/quarkus/deployment/pkg/NativeConfig.java [hunk 1]
```diff
@@ -77,13 +79,27 @@
...
+     * Defines the locales to be accessible in the native executable.
+     * The format is a comma separated list of IETF BCP 47 language tags,
+     * e.g. "fr,de,cs" or "de-AT,ja-JP" etc.
+     * See also LocalesBuildTimeConfig for a separate, native-image unrelated usage.
+     * <p>
+     * Defaults to empty, i.e. not explicitly set during native-image build.
+     */
+    @ConfigItem
+    @ConvertWith(TrimmedStringConverter.class)
+    public Optional<String> locales;
```
Comment (gsmet) [inline core/deployment/src/main/java/io/quarkus/deployment/pkg/NativeConfig.java:101]: Should we introduce specific locales for native while we have this one already: https://github.com/quarkusio/quarkus/blob/main/core/runtime/src/main/java/io/quarkus/runtime/LocalesBuildTimeConfig.java (and it needs to be set properly if you want the HV locale support work properly).

My guess is that we could also deprecate the two others that are already there (but still support them for now) and use `quarkus.default-locale` instead.

**Review by gsmet — CHANGES_REQUESTED** [review]
I added a comment on the config. Let's discuss it before adding one more config property.

**Review by gsmet — CHANGES_REQUESTED** [review]
Added some more comments on the code.

One last thing, I'm not sure if I have documented properly the existing config properties but if not, I think we should have a paragraph explaining these things in the native guide.

Because I'm not sure people will look to the config property table by default and this is rather fundamental if you hope to have a multi-locale app.

**Comment by gsmet** [comment]
Nice detective work, @Karm ! My comments are mostly cosmetic.

File: core/deployment/src/main/java/io/quarkus/deployment/steps/LocaleProcessor.java [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+    @BuildStep(onlyIf = { NativeBuild.class, NonEnglishLocale.class })
+    ReflectiveClassBuildItem setupReflectionClasses() {
+        return new ReflectiveClassBuildItem(false, false,
+                "sun.util.resources.provider.SupplementaryLocaleDataProvider",
+                "sun.util.resources.provider.LocaleDataProvider");
+    }
+
+    @BuildStep(onlyIf = { NativeBuild.class, NonEnglishLocale.class })
+    void servicesResource(BuildProducer<NativeImageResourceBuildItem> nibProducer,
+            BuildProducer<GeneratedResourceBuildItem> genResProducer) {
```
[the remark this answers is not in the document]
Comment (Karm) [reply core/deployment/src/main/java/io/quarkus/deployment/steps/LocaleProcessor.java:39]: Ack. Will do.

**Comment by Karm** [comment]
Thx @gsmet for the review.

**TL;DR:** I will try to get about without an additional config property and to make it nicer without blowing up Hibernate...

Ad 

 > I wouldn't use the regexp here, given things are properly split.
 > And if you use the config property I pointed out, it will be even better as
 > you will have proper locales.

 > This is extremely fragile, you should have used a List to get the data structured.
 > If you use the config properties I pointed out, it should be solved by itself.

I am aware of how it's done for `public Set<Locale> locales;` in [LocalesBuildTimeConfig.java](https://github.com/quarkusio/quarkus/blob/main/core/runtime/src/main/java/io/quarkus/runtime/LocalesBuildTimeConfig.java#L24)
and how it's converted in [LocaleConverter.java](https://github.com/quarkusio/quarkus/blob/main/core/runtime/src/main/java/io/quarkus/runtime/configuration/LocaleConverter.java#L25).

I don't know why it's happening though, or to put it better, why would I need to "deserialize" the comma separated list into a list (or set) of Locale objects only to "serialize" it again to a comma separated list of  IETF BCP 47 values to be fed to Graal's `H:IncludeLocales=`.  It seems to be all done (only?) for Hibernate.

In addition to that, I can see usages in e.g. `hibernate-orm-rest-data-panache` that rely on an incorrect representation with underscore: `quarkus.locales=en_US` that is happily accepted and normalized by [LocaleConverter.java](https://github.com/quarkusio/quarkus/blob/main/core/runtime/src/main/java/io/quarkus/runtime/configuration/LocaleConverter.java#L25).

I cannot feed an incorrect, underscore separated representation to Graal's `H:IncludeLocales=` as it would be silently ignored, i.e. `uk-UA` is correct, `uk_UA` is ignored.

> Should we introduce specific locales for native while we have this one already: [LocalesBuildTimeConfig.java](https://github.com/quarkusio/quarkus/blob/main/core/runtime/src/main/java/io/quarkus/runtime/LocalesBuildTimeConfig.java)
> (and it needs to be set properly if you want the HV locale support work properly).
> My guess is that we could also deprecate the two others that are already there (but still support them for now)
> and use quarkus.default-locale instead.

I introduced a separate property as I wanted to just make it work first, without delving into what Hibernate expects and refactoring existing properties or their semantics. Effectively jumping over the rabbit hole :) 

I will climb down none the less though and I will do another commit refactoring those locales related properties and show it to you again.
In a contemporary Mandrel/Graal native image, we need to care about:

1. Having a *default language*;
That could be covered by a Quarkus-wide `quarkus.user-language` instead of a native only `quarkus.native.user-language`. It will be translated for Graal's config as `-Duser.language=`.
2. Having a *default country* (affects also currency and timezone);
That could be covered by a Quarkus-wide `quarkus.user-country` instead of a native only `quarkus.native.user-country`. It will be translated for Graal's config as `-Duser.country=`.
3. Having a list of *additional* locales the native application is supposed to know about at runtime;
That could be covered by a single `quarkus.locales`, unifying what is currently done by me in `quarkus.native.locales` and what exists already in `quarkus.locales`.

I would drop a notion of "default locale", because that is deprecated in the contemporary Graal, to be replaced by currently used `user.language` and `user.country` options.

**Comment by gsmet** [comment]
I think, it's better to keep the config as I did it in the global config. I haven't done that only for HV. The whole point of it was to be used for other purposes next.

Having a proper locale makes sure you have a valid language - country association. These things are correlated.

You can push language and country to GraalVM by using the getters of the locale object.

As for your `Set` question, you relying on a regexp for:
```
 (nativeConfig.locales.isPresent() && !US_ENGLISH.matcher(nativeConfig.locales.get()).matches());
```
is a good example of why using a more structured approach is better. You shouldn't use a regexp, you should test the locales properly. And that is far easier with my approach.

With the existing config in https://github.com/quarkusio/quarkus/blob/main/core/runtime/src/main/java/io/quarkus/runtime/LocalesBuildTimeConfig.java , you can do everything you did in a cleaner and more structured way.

And you can push the info to GraalVM with very little adjustments.

And if we need the Locale(s) for other parts of the code, we have them in a structured way.

File: core/deployment/src/main/java/io/quarkus/deployment/pkg/NativeConfig.java (hunk 1 above)
Comment (Karm) [reply core/deployment/src/main/java/io/quarkus/deployment/pkg/NativeConfig.java]: Ack. Done.

File: core/deployment/src/main/java/io/quarkus/deployment/steps/LocaleProcessor.java [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+        public NonEnglishLocale(NativeConfig nativeConfig) {
+            this.nativeConfig = nativeConfig;
+        }
+
+        @Override
+        public boolean getAsBoolean() {
+            return (nativeConfig.userLanguage.isPresent() && !US_ENGLISH.matcher(nativeConfig.userLanguage.get()).matches())
+                    ||
+                    (nativeConfig.userCountry.isPresent() && !US_ENGLISH.matcher(nativeConfig.userCountry.get()).matches())
+                    ||
```
[the remark this answers is not in the document]
Comment (Karm) [reply core/deployment/src/main/java/io/quarkus/deployment/steps/LocaleProcessor.java]: Done.

File: core/deployment/src/main/java/io/quarkus/deployment/steps/LocaleProcessor.java [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+                    (nativeConfig.locales.isPresent() && !US_ENGLISH.matcher(nativeConfig.locales.get()).matches());
```
[the remark this answers is not in the document]
Comment (Karm) [reply core/deployment/src/main/java/io/quarkus/deployment/steps/LocaleProcessor.java]: Ack.

**Comment by Karm** [comment]
Hello @gsmet, 

I pushed a commit that does locales properties handling differently. I believe according to what you suggested.

Both `quarkus.native.user-language` and `quarkus.native.user-country` are deprecated so as users use the one global default locale setting `quarkus.default-locale`. These deprecated options take precedence though so as no user's config is broken.

Graal's `-Duser.language=` and `-Duser.country=` are derived from `quarkus.default-locale` and if present, from those deprecated `quarkus.native.user-language` and `quarkus.native.user-country` with those latter ones taking precedence.

The global `quarkus.locales` is used to construct Graal's additional locales for `-H:IncludeLocales=`.

If a locale is already present in any of  `quarkus.native.user-language`,  `quarkus.native.user-country`, `quarkus.default-locale`, and then again in `quarkus.locales`, it is not repeated in `-H:IncludeLocales=`. It captures only additional locales a it should.

I took a look at some Hibernate tests and they keep working with this change. Let me show some examples:

## hibernate-validator
Having only this and nothing else locales related in application.properties:
```
quarkus.locales=en,hr-HR,en-US,fr-FR
```
results in:
```
 -J-Duser.language=en -J-Duser.country=US -H:IncludeLocales=hr-HR,fr-FR
```
And that is expected. `en-US` is the default locale of my system and these `hr-HR,fr-FR` are correctly identified as additional.


## hibernate-orm-graphql-panache
Having only this and nothing else locales related in application.properties:
```
quarkus.default-locale=en_US
quarkus.locales=en_US
```
results in:
```
-J-Duser.language=en -J-Duser.country=US
```
And that is correct. There are no additional locales needed so the `-H:IncludeLocales` argument  is not set.

## locales (test suite in this PR)
Having only this and nothing else locales related in application.properties:
```
quarkus.locales=de,fr,ja,uk-UA,en
quarkus.native.user-language=cs
quarkus.default-locale=en-US
```
It correctly warns the user about a deprecated config property:
```
[WARNING] [io.quarkus.deployment.steps.LocaleProcessor]
Your application is setting the deprecated 'quarkus.native.user-language' configuration key.
Please, consider using only 'quarkus.default-locale' configuration key instead.
```
although it uses the property nonetheless, resulting in this correct setup:
```
-J-Duser.language=cs -J-Duser.country=US -H:IncludeLocales=de,fr,ja,uk-UA
```
Despite the global `quarkus.default-locale` being `en-US`, the deprecated `quarkus.native.user-language` overwrote its language to `cs`. Additional locales are correctly identified. Note that there is no `en` in `-H:IncludeLocales` despite the fact it is present in `quarkus.locales`. It was correctly removed as unnecessary additional locale.


When you have time, I will be glad for your feedback.

Cheers

File: core/deployment/src/main/java/io/quarkus/deployment/pkg/NativeConfig.java [hunk 5]
```diff
@@ -68,21 +70,29 @@
...
-    @ConvertWith(TrimmedStringConverter.class)
-    public Optional<String> userLanguage;
+    @ConfigItem
+    @ConvertWith(LocaleConverter.class)
+    @Deprecated
+    public Optional<Locale> userLanguage;
 
     /**
      * Defines the user country used for building the native executable.
+     * It also serves as te default Locale country for the native executable application runtime.
```
Comment (gsmet) [inline core/deployment/src/main/java/io/quarkus/deployment/pkg/NativeConfig.java:86]:
```suggestion
     * It also serves as the default Locale country for the native executable application runtime.
```

File: core/deployment/src/main/java/io/quarkus/deployment/steps/LocaleProcessor.java [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+                    ||
+                    nonEnglishLocale(localesBuildTimeConfig.defaultLocale)
+                    ||
+                    localesBuildTimeConfig.locales.stream().anyMatch(LocaleProcessor::nonEnglishLocale);
+        }
+    }
+
+    public static boolean nonEnglishLocale(Locale locale) {
+        return (!locale.getLanguage().isEmpty() && !"en".equals(locale.getLanguage()))
+                || (!locale.getCountry().isEmpty() && !"US".equals(locale.getCountry()));
```
Comment (gsmet) [inline core/deployment/src/main/java/io/quarkus/deployment/steps/LocaleProcessor.java:90]: `en_US` is always the default locale included in the native executable? My guess would have been that it's the default locale from your system instead (which happens to be `en_US` in your case).
So that this should be `nonDefaultLocale`. But I haven't looked at the GraalVM code so I might be incorrect.

File: core/deployment/src/main/java/io/quarkus/deployment/steps/LocaleProcessor.java [hunk 7]
```diff
@@ -0,0 +1,151 @@
...
+
+/**
+ * In order for a Native image built app to be able to use localized names of e.g. countries,
+ * these language bundles have to be loaded. JDK uses ServiceLoader approach for that.
+ * For instance, Locale.FRANCE.getDisplayCountry(Locale.GERMAN) must print "Frankreich".
+ */
+public class LocaleProcessor {
+
+    private static final Logger log = Logger.getLogger(LocaleProcessor.class);
+    public static final String DEPRECATED_USER_LANGUAGE_WARNING = "Your application is setting the deprecated 'quarkus.native.user-language' configuration key. "
```
Comment (gsmet) [inline core/deployment/src/main/java/io/quarkus/deployment/steps/LocaleProcessor.java:30]:
```suggestion
    public static final String DEPRECATED_USER_LANGUAGE_WARNING = "Your application is setting the deprecated 'quarkus.native.user-language' configuration property. "
```

File: core/deployment/src/main/java/io/quarkus/deployment/steps/LocaleProcessor.java [hunk 8]
```diff
@@ -0,0 +1,151 @@
...
+            +
+            "Please, consider using only 'quarkus.default-locale' configuration key instead.";
```
Comment (gsmet) [inline core/deployment/src/main/java/io/quarkus/deployment/steps/LocaleProcessor.java:32]:
```suggestion
            "Please, consider using only 'quarkus.default-locale' configuration property instead.";
```

File: core/deployment/src/main/java/io/quarkus/deployment/steps/LocaleProcessor.java [hunk 9]
```diff
@@ -0,0 +1,151 @@
...
+    public static final String DEPRECATED_USER_COUNTRY_WARNING = "Your application is setting the deprecated 'quarkus.native.user-country' configuration key. "
```
Comment (gsmet) [inline core/deployment/src/main/java/io/quarkus/deployment/steps/LocaleProcessor.java:33]:
```suggestion
    public static final String DEPRECATED_USER_COUNTRY_WARNING = "Your application is setting the deprecated 'quarkus.native.user-country' configuration property. "
```

File: core/deployment/src/main/java/io/quarkus/deployment/steps/LocaleProcessor.java [hunk 10]
```diff
@@ -0,0 +1,151 @@
...
+            +
+            "Please, consider using only 'quarkus.default-locale' configuration key instead.";
```
Comment (gsmet) [inline core/deployment/src/main/java/io/quarkus/deployment/steps/LocaleProcessor.java:35]:
```suggestion
            "Please, consider using only 'quarkus.default-locale' configuration property instead.";
```

File: core/deployment/src/main/java/io/quarkus/deployment/pkg/NativeConfig.java [hunk 11]
```diff
@@ -68,21 +70,29 @@
...
+     * 
+     * @deprecated Use the global quarkus.default-locale.
      */
-    @ConfigItem(defaultValue = "${user.country:}")
-    @ConvertWith(TrimmedStringConverter.class)
-    public Optional<String> userCountry;
+    @ConfigItem
+    @ConvertWith(LocaleConverter.class)
+    @Deprecated
+    public Optional<Locale> userCountry;
```
Comment (gsmet) [inline core/deployment/src/main/java/io/quarkus/deployment/pkg/NativeConfig.java:95]: How can you make this thing a Locale? It is not, it's just a country.

File: core/deployment/src/main/java/io/quarkus/deployment/pkg/NativeConfig.java [hunk 12]
```diff
@@ -68,21 +70,29 @@
...
+     * 
+     * @deprecated Use the global quarkus.default-locale.
      */
-    @ConfigItem(defaultValue = "${user.language:}")
-    @ConvertWith(TrimmedStringConverter.class)
-    public Optional<String> userLanguage;
+    @ConfigItem
+    @ConvertWith(LocaleConverter.class)
+    @Deprecated
+    public Optional<Locale> userLanguage;
```
Comment (gsmet) [inline core/deployment/src/main/java/io/quarkus/deployment/pkg/NativeConfig.java:82]: I think you should keep this thing a String as it's just the language part. Also see in another comment how it gets confusing.

**Review by gsmet — CHANGES_REQUESTED** [review]
Thanks! Definitely looks like what I had in mind. I spotted a few issues, some very minor, some not so much.

**Comment by Karm** [comment]
@gsmet Thx so much for spending the time explaining it :+1:  The latest commit hopefully makes it all fine.

## Typos / Javadoc

Fixed.

## en_US vs "nonDefaultLocale"

> `en_US` is always the default locale included in the native executable?
> My guess would have been that it's the default locale from your system
> instead (which happens to be en_US in your case).
> So that this should be nonDefaultLocale.
> But I haven't looked at the GraalVM code so I might be incorrect.

I though I was careful enough, but I wasn't. It was indeed incorrect.
While `en_US` is *always* present in a JVM implementation as mandated by:

 > The returned array represents the union of locales supported by the Java
 > runtime environment and by installed LocaleServiceProvider implementations.
 > It must contain at least a Locale instance equal to Locale.US.
Source: https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/util/Locale.html#getAvailableLocales()

It is (`en_US`) by no means always included in the Native image:

```
Locale.US.getDisplayCountry(Locale.US) throws:

Exception in thread "main" java.util.MissingResourceException:
  Can't find bundle for base name sun.util.resources.LocaleNames, locale en_US
```
when the default system locale is e.g.:
```
$ localectl status
   System Locale: LANG=cs_CZ.UTF-8
       VC Keymap: us
      X11 Layout: us
```

It is done by this substitution in Graal: [Target_java_util_ResourceBundle.java](https://github.com/oracle/graal/blob/master/substratevm/src/com.oracle.svm.core/src/com/oracle/svm/core/jdk/localization/substitutions/Target_java_util_ResourceBundle.java#L52) that bakes in the image whatever was `Locale.getDefault()` at the time.
The default is constructed in JDK of whatever is available in sys props: [Locale.java](https://github.com/openjdk/jdk11u-dev/blob/jdk-11.0.14+8/src/java.base/share/classes/java/util/Locale.java#L935).

Fun fact is that on a system that has really bare minimum, only 1 locale installed system wide and set to:

```
$ localectl status
   System Locale: LANG=C.UTF-8
```
It defaults to US in JDK and thus `Locale.getDefault()` is US in native-image and the aforementioned example works:

```
Locale.US.getDisplayCountry(Locale.US)):

United States
```

I fixed it in this PR by removing the incorrect `nonEnglishLocale` and replacing it with `nonDefaultLocale`.


## Country is not a Locale

> How can you make this thing a Locale? It is not, it's just a country.

I cannot apparently. While a BCP 47 language tag is enough to form a Locale, the ISO 3166 alpha-2 country code alone is not.
It "worked" for me, because even if you haven't set anything, the language was "en".

Fixed.

## String, Locale, back to String again

> I think you should keep this thing a String as it's just the language part.
> Also see in another comment how it gets confusing.

I agree it's better left as String.

# Removing superfluous locales

> You're making the country a locale and removing it from the list.
> You should create ONE locale from both the language and country and then remove it from the list.

I refactored to make it so: If user configured both legacy properties, one locale is formed and removed.
If there is at least language, a locale is formed and removed.

```
-    nativeConfig.userCountry.ifPresent(additionalLocales::remove);
-    nativeConfig.userLanguage.ifPresent(additionalLocales::remove);
+    if (nativeConfig.userCountry.isPresent() && nativeConfig.userLanguage.isPresent()) {
+        additionalLocales.remove(new Locale(nativeConfig.userLanguage.get(), nativeConfig.userCountry.get()));
+    } else if (nativeConfig.userLanguage.isPresent()) {
+        additionalLocales.remove(new Locale(nativeConfig.userLanguage.get()));
+    }

```

It is indeed possible to beat it by intentional misconfiguration of legacy properties...

**Review by gsmet — APPROVED** [review]
Looks good, thanks!

I actually added the ITs to the integration tests matrix. Let's see what CI has to say.

**Comment by Karm** [comment]
@gsmet  Opinion on Gradle Windows fail?

**Comment by gsmet** [comment]
Definitely unrelated!
